arrow-fx-stm/src/main/kotlin/arrow/fx/stm/TSet.kt

```diff
@@ -8,6 +8,97 @@ fun <A> STM.newTSet(fn: (A) -> Int): TSet<A> = TSet(newHamt(), fn)
 
+/**
+ *
+ * A [TSet] is a concurrent transactional implementation of a hashset.
+ *
+ * Based on a Hash-Array-Mapped-Trie implementation. While this does mean that a read may take up to 5 steps to be resolved (depending on how
+ *  well distributed the hash function is), it also means that structural changes can be isolated and thus do not increase contention with
+ *  other transactions. This effectively means concurrent access to different values is unlikely to interfere with each other.
+ *
+ * > Hash conflicts are resolved by chaining.
+ *
+ * ## Creating a [TSet]
+ *
+ * Depending on whether or not you are in a transaction you can use either [STM.newTSet] or [TSet.new] to create a new [TSet].
+ *
+ * There are a few alternatives because [TSet] can be supplied a custom hash strategy. If no argument is given it defaults to [hashCode].
+ *
+ * ## Adding elements to the set
+ *
+ * Adding an element can be achieved by using either [STM.insert] or its alias [STM.plus]:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TSet
+ * import arrow.fx.stm.atomically
+ * import arrow.fx.coroutines.Environment
+ *
+ * suspend fun main() {
+ *   Environment().unsafeRunSync {
+ *     //sampleStart
+ *     val tset = TSet.new<String>()
+ *     atomically {
+ *       tset.insert("Hello")
+ *       tset + "World"
+ *     }
+ *     //sampleEnd
+ *   }
+ * }
+ * ```
+ *
+ * ## Removing an element from the set
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TSet
+ * import arrow.fx.stm.atomically
+ * import arrow.fx.coroutines.Environment
+ *
+ * suspend fun main() {
+ *   Environment().unsafeRunSync {
+ *     //sampleStart
+ *     val tset = TSet.new<String>()
+ *     atomically {
+ *       tset.insert("Hello")
+ *       tset.remove("Hello")
+ *     }
+ *     //sampleEnd
+ *   }
+ * }
+ * ```
+ *
+ * ## Checking for membership
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.stm.TSet
+ * import arrow.fx.stm.atomically
+ * import arrow.fx.coroutines.Environment
+ *
+ * suspend fun main() {
+ *   Environment().unsafeRunSync {
+ *     //sampleStart
+ *     val tset = TSet.new<String>()
+ *     val result = atomically {
+ *       tset.insert("Hello")
+ *       tset.member("Hello")
+ *     }
+ *     //sampleEnd
+ *     println("Result $result")
+ *   }
+ * }
+ * ```
+ *
+ * ## Where are operations like `isEmpty` or `size`?
+ *
+ * This is a design tradeoff. It is entirely possible to track size however this usually requires one additional [TVar] for size and
+ *  almost every operation would modify that. That will lead to contention and thus decrease performance.
+ *
+ * Should this feature interest you and performance is not as important please open an issue. It is most certainly possible to add another version
+ *  of [TSet] that keeps track of its size.
+ *
+ */
+// Why a `Pair<Unit, A>`? Well because kotlin. Altering a Hamt is done through inline *reified* methods and because we cannot get
```

Comment:
I'm not sure I understand why `Pair<Unit, A>` instead of just `A`, also not from looking at the inline *reified* methods. Also 😂

Reply:
Iirc `reified` needs some proof of a concrete class to compile otherwise the calling methods must be reified and thus inline as well. That is not possible on interfaces because.... well idk. Either way `Pair<Unit, A>` seems to fulfill this need for a concrete class and thus avoids having to make the interface methods inline.

Maybe there are better solutions, maybe even inline classes to get rid of the wrapper after the fact...

Reply:
🤔 Not sure I understand what the difference is between `Hamt<A>` and `Hamt<Pair<Unit, A>>`, since it's wrapped in another class?

Reply:
Ah this is not about `Hamt<A>` not working, it is about `alterHamtWithHash` using methods like `arrayOf` which are `inline` and have `reified` generics. Because the elements used here are generic this requires `alterHamtWithHash` to be `inline` and use `reified` as well. The methods on `TSet` are implemented on the `STM` interface and use `alterHamtWithHash` and with a generic `A` as well. This means they would also need to be `inline` and use `reififed` but since we are on an interface that does not work. The solution is to supply `alterHamtWithHash` with a concrete class for the generic (`Pair` in this case) so that we don't need to propagate `inline` to the methods.

I should look into using an inline class here tho but iirc an inline class over generics may not disappear after compilation.

This sucks.

Reply:
Oh, I see now.. Thanks for the clarification! Not sure if `inline class` would help there, since I still expect it to do boxing.
The only other option I see is to use casting inside of `alterHamtWithHash` to remove the `reified` restriction on it.

use `arrayOf<Any?>(newEl, *branch.value) as Array<A>`, instead of `arrayOf<A>(newEl, *branch.value)` and you'll be able to get rid of `reified`. Not a must, and definitely not in this PR. It's a trick I used in `Chunk` in a couple of places as well to avoid `reified`.

Reply:
Ah that is a good idea! Will note that somewhere so I remember^^